review: conservative_continuous_gaussian_tail_to_alpha

Shoeboxam opened the pull request:
Closes #2060

In the "Files changed" diffs I added a space to the end of each lines to make it possible to leave comments inline. Either focus on leaving comments in the green, or comment on the pdf rendered below and upload it in your review. Thanks!

Shoeboxam committed:
review: conservative_continuous_gaussian_tail_to_alpha

Shoeboxam commented:
* **#2125** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

annabelle400 reviewed (commented):
The implementation for this relies on the implementation of the  erf function, together with rounding.

The proof consists of definitions and a lemma statement that says this.
There is included code that checks these assumptions which looks good!

Other than that, it would be good to match up the statement of the postcondition:

Should it read...
Pr[x>= t] ?

but perhaps because of the definition of t with rounding, it should be 
a strict inequality. Please could you confirm.

Also "at least" is easier to read than "no smaller than".

tudorcebere reviewed (commented):
The proof looks good! Left some small comments, but the corectness shouldn't change.

Shoeboxam committed:
respond to review 4/20

Shoeboxam commented:
Thanks for the reviews! Changes are in the latest commit (there's two commits now).

@annabelle400 
The function’s contract is intentionally kept as a bound on Pr[X > tail] for the ideal continuous Gaussian reference distribution. The thresholded discrete gaussian noise privacy map that calls it needs $\Pr[Z_j > T - Δ_∞]$. I use the tail bound for the continuous distribution as a (very slightly) conservative overestimation for the discrete gaussian tail bound when scale is large because the discrete tail bound isn't tractably computable in that regime. A guarantee on $\Pr[X \ge tail]$ would actually be different, even though we're talking about a continuous distribution, due to the float discretization.

In the proof I use $\Pr[X > tail] = \Pr[X >= tail]$ only as an internal step because that reference distribution is continuous; I’m not making that claim about the float implementation or any discretized/postprocessed sampler.

I added a comment in the proof to this effect, it's confusingly specific otherwise and I had forgotten about this rationale myself!

@tudorcebere 
I updated the proof to restate everything in terms of scale/tail, fixed the sign in the erf step, avoided reusing the integration variable, and added the missing return in the pseudocode.
